### PR TITLE
[SPARK-41319][CONNECT][PYTHON] Implement Column.{when, otherwise} and Function `when` with `UnresolvedFunction`

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?

1, Implement `Column.{when, otherwise}` and Function `when`

### Why are the changes needed?
For API coverage

### Does this PR introduce _any_ user-facing change?
yes

### How was this patch tested?
added UT

### CODE REVIEW

File: python/pyspark/sql/connect/column.py [hunk 1]
```diff
@@ -129,6 +140,53 @@ def name(self) -> str:
...
+        for condition, value in self._branches:
+            args.append(condition)
+            args.append(value)
+
+        if self._else_value is not None:
+            args.append(self._else_value)
+
+        unresolved_function = UnresolvedFunction(name="when", args=args)
+
+        return unresolved_function.to_plan(session)
```
[the remark this answers is not in the document]
Comment (zhengruifeng) [reply python/pyspark/sql/connect/column.py:182]: @hvanhovell  @amaliujia

[the remark this answers is not in the document]
Comment (zhengruifeng) [reply python/pyspark/sql/connect/column.py:182]: let me add another constructor

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala [hunk 2]
```diff
@@ -161,6 +161,15 @@ case class CaseWhen(
     elseValue: Option[Expression] = None)
   extends ComplexTypeMergingExpression with ConditionalExpression {
 
+  def this(branches: Seq[Expression]) = this(
```
Comment (zhengruifeng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:164]: still fail with:
```
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "[FAILED_FUNCTION_CALL] Failed preparing of the function `when` for call. Please, double check function's arguments."
        debug_error_string = "{"created":"@1670384303.828721000","description":"Error received from peer ipv6:[::1]:15002","file":"src/core/lib/surface/call.cc","file_line":1064,"grpc_message":"[FAILED_FUNCTION_CALL] Failed preparing of the function `when` for call. Please, double check function's arguments.","grpc_status":2}"
>
```

**Comment by zhengruifeng** [comment]
it keep failing in the invocation of constructor, I print the constructor as well as expressions in the error message, and they looks fine

```
grpc._channel._MultiThreadedRendezvous: <_MultiThreadedRendezvous of RPC that terminated with:
        status = StatusCode.UNKNOWN
        details = "[FAILED_FUNCTION_CALL] Failed preparing of the function `when_1, constructor:public org_apache_spark_sql_catalyst_expressions_CaseWhen(scala_collection_Seq), exprs:class org_apache_spark_sql_catalyst_expressions_EqualTo:(cast(a#6 as bigint) = 0) , class org_apache_spark_sql_catalyst_expressions_Literal:1_0 , class org_apache_spark_sql_catalyst_expressions_Literal:2_0, error:null` for call. Please, double check function's arguments."
        debug_error_string = "{"created":"@1670408082.289473000","description":"Error received from peer ipv6:[::1]:15002","file":"src/core/lib/surface/call.cc","file_line":1064,"grpc_message":"[FAILED_FUNCTION_CALL] Failed preparing of the function `when_1, constructor:public org_apache_spark_sql_catalyst_expressions_CaseWhen(scala_collection_Seq), exprs:class org_apache_spark_sql_catalyst_expressions_EqualTo:(cast(a#6 as bigint) = 0) , class org_apache_spark_sql_catalyst_expressions_Literal:1_0 , class org_apache_spark_sql_catalyst_expressions_Literal:2_0, error:null` for call. Please, double check function's arguments.","grpc_status":2}"
```

I also check in the shell:
```
scala> import org.apache.spark.sql.catalyst.expressions._
import org.apache.spark.sql.catalyst.expressions._

scala> val casewhen = CaseWhen(Seq.empty, None)
casewhen: org.apache.spark.sql.catalyst.expressions.CaseWhen = CASE END

scala> val varargCtor = casewhen.getClass.getConstructors().find(_.getParameterTypes.toSeq == Seq(classOf[Seq[_]]))
varargCtor: Option[java.lang.reflect.Constructor[_]] = Some(public org.apache.spark.sql.catalyst.expressions.CaseWhen(scala.collection.Seq))

scala> val expressions = Seq((col("a").cast("long") === 0).expr, lit(1.0).expr, lit(2.0).expr)
expressions: Seq[org.apache.spark.sql.catalyst.expressions.Expression] = List((cast('a as bigint) = 0), 1.0, 2.0)

scala> varargCtor.get.newInstance(expressions)
res10: Any = CASE WHEN (cast('a as bigint) = 0) THEN 1.0 ELSE 2.0 END
```

**Comment by zhengruifeng** [comment]
Fixed, the root cause is the previous constructor copied from https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala#L388-L395  failed due to `MatchError` ...

**Comment by zhengruifeng** [comment]
also cc @HyukjinKwon @cloud-fan @grundprinzip @amaliujia @xinrong-meng

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala [hunk 3]
```diff
@@ -161,6 +161,13 @@ case class CaseWhen(
...
```
Comment (grundprinzip) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:164]: Does this need an additional test?

**Review by grundprinzip — APPROVED** [review]
LGTM with one comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala (hunk 3 above)
Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:164]: let me see

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:164]: Can we use the `CaseWhen.createFromParser`? You will need to register an `ExpressionBuilder` for this. The upside of this is that it is less messy in the constructor, and that there is ample testing for this from the parser.

Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:164]: @hvanhovell  what about we create CaseWhen in the [connect planner ](https://github.com/apache/spark/blob/master/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala#L535-L554), then we don't need to change sql side at all.

Comment (hvanhovell) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:164]: I think it is better to keep the connect planner as generic as possible. If we can use an unresolved function then it is better.

Comment (zhengruifeng) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala:164]: oh, I mean we still use an unresolved function, and we can call `CaseWhen.createFromParser` in connect planner

**Comment by zhengruifeng** [comment]
cc @HyukjinKwon @cloud-fan @amaliujia

**Review by amaliujia — APPROVED** [review]
LGTM

**Comment by zhengruifeng** [comment]
merged into master, thanks for the reviews

**Review by HyukjinKwon — COMMENTED** [review]
LGTM, thx.

**Pull request closed without merging**